### PR TITLE
Deterministic version of CUDA forces and stresses kernels

### PR DESCRIPTION
**Pull request by mtaillefumier**
Calculations of the forces and stress are deterministic on GPU. It does not imply that the DeepMD code is deterministic by default as TensorFlow also requires to be set up properly either at runtime or during the initialization phase. 

To obtain the same model parameters, add the following variables to the job scripts

```[sh]
export TF_DETERMINISTIC_OPS=1
export TF_INTER_OP_PARALLELISM_THREADS=0
export TF_INTRA_OP_PARALLELISM_THREADS=0
```
Details of the changes:

- Remove the use of atomic operations in the forces and stress kernels.
- Use template programming to minimize code duplication and minor refactoring

Authors :
- (@asedova) A. Sedova (ORNL) 
- (@mtaillefumier) M. Taillefumier (ETH Zurich / CSCS) 

See PR #3684 for comments

## Summary by CodeRabbit

- **New Features**
  - Enhanced force and virial calculations in molecular dynamics simulations for improved accuracy and performance.
  - Added a new documentation guide on achieving determinism in DeepMD simulations.

- **Refactor**
  - Optimized memory access and thread synchronization in GPU calculations.

- **Documentation**
  - Provided detailed guidance on configuring deterministic runs in DeepMD using TensorFlow settings.

- **Bug Fixes**
  - Removed deprecated functions to streamline codebase and enhance maintainability.

### CODE REVIEW

**Comment by njzjz** [comment]
It seems that the current change cannot make the GPU CI pass.

**Comment by mtaillefumier** [comment]
> It seems that the current change cannot make the GPU CI pass.

The errors are unclear to me. It is not a compilation problem (which I would have fixed asap) but something deep in python. One file is just a text file, the two other files are source code. The there is no change in the API, no memory leak (I checked for this), etc.... I can also compile deepmd in containers etc.... 

NB : I use this docker file 

```
FROM tensorflow/tensorflow:2.16.1-gpu

CMD ["/bin/bash"]
RUN apt-get update -qq && apt-get install -qq --no-install-recommends \
    cmake \
    gcc-12 \
    g++-12 \
    gfortran-12 \
    git
ENV DP_VARIANT=cuda

RUN git clone https://github.com/mtaillefumier/deepmd-kit.git
git checkout devel
WORKDIR deepmd-kit
ENV DP_VARIANT=cuda

RUN pip install .
ENTRYPOINT []
```

**Comment by mtaillefumier** [comment]
Something is wrong with this patch. It works with master but fails with the devel branch. I am missing something.

**Comment by mtaillefumier** [comment]
Sorry for the delay. I think I found the root of the problem. A missing synchronization barrier.

File: source/lib/src/gpu/prod_virial.cu [hunk 1]
```diff
@@ -26,112 +26,140 @@ __global__ void atom_virial_reduction(FPTYPE* virial,
...
-      virial, atom_virial, net_deriv, in_deriv, rij, nlist, nloc, nnei);
+  dim3 block_grid = dim3(nloc, 1, 1);
+  // to accomodate AMD warp size
+  dim3 thread_grid = dim3(64, 1, 1);
+  virial_deriv_wrt_neighbors<FPTYPE, radial_only_, 64>
+      <<<block_grid, thread_grid>>>(virial, atom_virial, net_deriv, in_deriv,
+                                    rij, nlist, nloc, nnei);
   DPErrcheck(gpuGetLastError());
   DPErrcheck(gpuDeviceSynchronize());
   // reduction atom_virial to virial
```
[the remark this answers is not in the document]
Comment (mtaillefumier) [reply source/lib/src/gpu/prod_virial.cu:144]: irrelevant tt is **always** lower than blockDim.x

**Comment by mtaillefumier** [comment]
I found why this PR fails. I missed something important when I modified these kernels and I know how to correct them now. I will close the PR for the moment so that I can push changes without triggering the CI/CD all the time.

**Pull request closed without merging**